src/docs/binding-closures.md

```diff
@@ -0,0 +1,135 @@
+# Binding Closures
+
+Closures in Swift come in two general forms: escaping and non-escaping. A closure is considered escaping if it consinues to exist outside of the context that created it. For example, a closure paramater to a function that gets stored in a member variable in a class *must* be marked `@escaping` or the compiler will flag it as an error.
+
+In either case, Swift closures, unlike some other languages, are full-fledged closures that can have both free and bound variables within the body of the closure.
+
+The distinction between the two is that escaping closures that capture free variables use a dynamically allocated object to contain any captured variables. This object is reference counted. Non-escaping closures use stack allocated memory to contain any captured variables.
+
+Beyond the broad classification of escaping/non-escaping, closures can also be async and they can throw.
+
+Internally, closures are represented as 2 machine words:
+- Pointer to a function entry point for the closure (see below for details)
+- Pointer to the data context object for captured free variables or 0/null if there is no context object
```

Comment:
Is this passed through context register?

Reply:
Yes.

Reply:
The data context object implicitly added and not displayed in the Swift signature but can be seen in LLVM IR, right? I remember we were hitting issues with callbacks where an implicit pointer like argument was being added after the function pointer. Here are the runtime tests for callbacks https://github.com/dotnet/runtime/blob/main/src/tests/Interop/Swift/SwiftCallbackAbiStress/SwiftCallbackAbiStress.cs